util/types/time.go

```diff
@@ -1759,25 +1760,47 @@ func abbrDayOfMonth(day int) string {
 // StrToDate converts date string according to format.
 // See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_date-format
 func (t *Time) StrToDate(date, format string) bool {
-	var tm mysqlTime
+	ctx := make(map[string]int)
+	tm := mysqlTimeEx{context: ctx}
 	if !strToDate(&tm, date, format) {
 		t.Time = ZeroTime
 		t.Type = mysql.TypeDatetime
 		t.Fsp = 0
 		return false
 	}
+	mysqlTimeExFix(&tm)
 
-	t.Time = tm
+	t.Time = tm.mysqlTime
 	t.Type = mysql.TypeDatetime
 	if t.check() != nil {
 		return false
 	}
 	return true
 }
 
+// mysqlTimeEx wraps around mysqlTime and provides the context for parsing.
+// Some date format parser is not context free, for example, day of year needs
+// to know which year, it can't be interpreted alone.
+type mysqlTimeEx struct {
+	mysqlTime
+	context map[string]int
+}
+
+func mysqlTimeExFix(t *mysqlTimeEx) {
+	ctx := t.context
+	if yearOfDay, ok := ctx["%j"]; ok {
+		_ = yearOfDay // TODO
```

Comment:
It rely on some function do the reverse work of `Daynr`, to compute the date according to the day in a year,  but I found we haven't implement yet, and that would be another PR.